### ISSUE TITLE
autocomplete will incorrectly apply expressions in CJK language IME

### ISSUE THREAD

**Issue by XHXIAIEIN** [Scirra/Construct-bugs#7988]
## Attach a .c3p

[issue-autocomplete-ime.c3p.zip](https://github.com/Scirra/Construct-bugs/files/15223340/issue-autocomplete-ime.c3p.zip)

## Steps to reproduce
1. input `Sprite.` display the autocomplete list.

<img width="318" alt="Snipaste_2024-05-07_00-12-00" src="https://github.com/Scirra/Construct-bugs/assets/45864744/bd07db1a-164d-4767-8851-f2a3958cb2d8">

2. input `序号` (xuhao) use IME. note: You can directly paste the `序号` into the input box. This operation does not affect the reproduction results.

<img width="302" alt="Snipaste_2024-05-07_00-12-35" src="https://github.com/Scirra/Construct-bugs/assets/45864744/cdad8405-564b-43c5-a999-26a37f5e1943">

3. Note that autocomplete has not been updated at this time.  If any key is pressed in this state( `Enter` or `+` or `-`), it will cause the first expression in the selection list to be automatically completed.

<img width="243" alt="Snipaste_2024-05-07_00-16-08" src="https://github.com/Scirra/Construct-bugs/assets/45864744/347de821-7bfb-4df5-bf9b-c262f7f991fa">

## Observed result

Autocomplete will automatically apply the first one in the expression 
list. Note that this problem only occurs when the autocomplete list appears.

Or in the process of input the `.` this expression has been selected in the autocomplete list, then it will work normally.

## Expected result

works as expected.

## More details

consider referencing some third-party `pinyin` libraries and applying them when the editor selects CJK language.  Also, for poeditor suggestions, please refer to this:
https://github.com/Scirra/Construct-feature-requests/issues/95

**Affected browsers/platforms:** 

**First affected release:** 

## System details

<details><summary>View details</summary>

PASTE HERE

</details>

**Comment by XHXIAIEIN** [comment]
I don't know if this helps, maybe this can be used to solve the problem:
https://developer.mozilla.org/en-US/docs/Web/API/Element/compositionstart_event

*Observed result*

你好

```
compositionstart: 
compositionupdate: n
compositionupdate: ni
compositionupdate: ni'h
compositionupdate: ni'ha
compositionupdate: ni'hao
compositionupdate: 你好
**compositionend**: 你好
```

**Comment by imKiBi** [comment]
I have long been troubled by similar issues unrelated to Chinese or IME. 
![6FHPKIT %(T92A)EGXEX59X](https://github.com/Scirra/Construct-bugs/assets/96277390/279fde38-9b29-4134-a166-a6b5f3799942)
When I type "&newline&", a selection box highlighting my "Name....." item pops up, which is normal. 
However, even after typing "&newline" and it turning blue, the system still automatically rewrites to me when I type in an “&”. 
![{S6{AAYZ_0P OOWE3X$BAQM](https://github.com/Scirra/Construct-bugs/assets/96277390/7590fa90-6881-40b8-a30b-8e2aa2180a13)
Please refer to the pic ↑
Hope C3 can do better and better :)

**Comment by r2680143448** [comment]
I've also been experiencing this issue, where searching with Chinese input is problematic. I really hope the official team can address this matter.

![6969](https://github.com/user-attachments/assets/2dcfab87-0ef6-44d6-b017-f6713c550e3d)

![6969](https://github.com/user-attachments/assets/b3e76ad9-ebcb-4bb5-9d39-0458af27371f)

**Comment by AshleyScirra** [comment]
This is tricky to test because unfortunately I'm not familiar with Chinese or entering text via IME. However it looks like the Windows emoji picker (opened with `Windows` + `.`) is a type of IME entry which also fires composition events. So it seems this may be a good way to test it, using emoji in place of Chinese characters: so long as it autocompletes correctly with the emoji picker, then I think that means IME should work with other languages too.

I was able to use the emoji-picker approach to reproduce two variants of this issue: the originally noted one involving typing `Sprite.` followed by IME entry, and another using IME to refer to an object name (i.e. global namespace autocomplete). Hopefully this solves the problem, but I'm marking this issue as to check in the next release so others can verify how it works with IME entry for other languages like Chinese.

**Comment by AshleyScirra** [comment]
r405 is now out with improvements for IME. Does it work better there now?

**Comment by XHXIAIEIN** [comment]
Yes, it is work better now! I gave it a quick try and it did solve the problem, thanks you!

**Comment by XHXIAIEIN** [comment]
There is another issue:
Asian symbols do not trigger autocompletion
> Expressions can now use Asian characters as equivalents for latin operators: ，。“”（）？：

**Comment by XHXIAIEIN** [comment]
other issue:
In expression, the IME will inserts the first character.

![image](https://github.com/user-attachments/assets/a4c48bcc-b554-4b90-986d-3bf12dfe789e)

## Steps to reproduce:
1. abs()|
2. abs(|)
3. abs(jingling|)
4. abs(j精灵|)

**Comment by AshleyScirra** [comment]
I believe the issue described in the original report is now fixed. If there are other issues to follow up it would be best to file them separately.
